namesys/pubsub.go

```diff
@@ -0,0 +1,346 @@
+	}
+
+	// the datastore is shared with the routing publisher to properly increment and persist
+	// ipns record sequence numbers; so we need to Record our new entry in the datastore
+	dsrec, err := record.MakePutRecord(k, ipnskey, data, true)
+	if err != nil {
+		return err
+	}
+
+	dsdata, err := proto.Marshal(dsrec)
+	if err != nil {
+		return err
+	}
+
+	p.ds.Put(dshelp.NewKeyFromBinary([]byte(ipnskey)), dsdata)
+
+	// now we publish, but we also need to bootstrap pubsub for our messages to propagate
+	topic := "/ipns/" + ID.Pretty()
+
+	p.mx.Lock()
+	_, ok := p.subs[topic]
+
+	if !ok {
+		p.subs[topic] = struct{}{}
+		p.mx.Unlock()
+
+		bootstrapPubsub(p.ctx, p.cr, p.host, topic)
+	} else {
+		p.mx.Unlock()
+	}
+
+	log.Debugf("PubsubPublish: publish IPNS record for %s", topic)
+	return p.ps.Publish(topic, data)
+}
+
+// Resolve resolves a name through pubsub and default depth limit
+func (r *PubsubResolver) Resolve(ctx context.Context, name string) (value path.Path, err error) {
+	return r.ResolveN(ctx, name, DefaultDepthLimit)
+}
+
+// ResolveN resolves a name through pubsub with the specified depth limit
+func (r *PubsubResolver) ResolveN(ctx context.Context, name string, depth int) (value path.Path, err error) {
+	return resolve(ctx, r, name, depth, "/ipns/")
+}
+
+func (r *PubsubResolver) resolveOnce(ctx context.Context, name string) (value path.Path, err error) {
+	log.Debugf("PubsubResolve: '%s'", name)
+
+	// retrieve the public key once (for verifying messages)
+	xname := strings.TrimPrefix(name, "/ipns/")
+	hash, err := mh.FromB58String(xname)
+	if err != nil {
+		log.Warningf("PubsubResolve: bad input hash: [%s]", xname)
+		return
+	}
+
+	pubk, err := r.pkf.GetPublicKey(ctx, peer.ID(hash))
+	if err != nil {
+		log.Warningf("PubsubResolve: error fetching public key: %s [%s]", err.Error(), xname)
+		return
+	}
+
+	// the mutex is locked both for subscription map manipulation and datastore
+	// manipulation (otherwise race in TTL checking/invalidation)
+	r.mx.Lock()
+	defer r.mx.Unlock()
+
+	// see if we already have a pubsub subscription; if not, subscribe
+	sub, ok := r.subs[name]
+	if !ok {
+		sub, err = r.ps.Subscribe(name)
+		if err != nil {
+			return
+		}
+
+		r.subs[name] = sub
+		go r.handleSubscription(sub, name, pubk)
+		go bootstrapPubsub(r.ctx, r.cr, r.host, name)
+	}
+
+	// resolve to what we may already have in the datastore
+	dsval, err := r.ds.Get(dshelp.NewKeyFromBinary([]byte(name)))
+	if err != nil {
+		// this signals ds.ErrNotFound when we have no record
+		return
+	}
+
+	entry := dsval.(*pb.IpnsEntry)
+
+	// check EOL; if the entry has expired, delete from datastore and return ds.ErrNotFound
+	eol, ok := checkEOL(entry)
+	if ok && eol.Before(time.Now()) {
+		r.ds.Delete(dshelp.NewKeyFromBinary([]byte(name)))
+		return "", ds.ErrNotFound
+	}
+
+	value, err = path.ParsePath(string(entry.GetValue()))
+	return
+}
+
+func (r *PubsubResolver) handleSubscription(sub *floodsub.Subscription, name string, pubk ci.PubKey) {
```

Comment:
Added a PubsubResolver.Cancel method in https://github.com/ipfs/go-ipfs/pull/4047/commits/176fcf185b8ed450d2430522cd3dbb3620192e51